ibm_mq/tests/test_queue_metric_collector.py

```diff
@@ -35,3 +38,241 @@ def test_regex_precedes_autodiscovery(instance):
+            if auto_discover_queues_via_names:
+                assert any(
+                    "Error inquiring queue names for pattern" in record.message
+                    for record in caplog.records
+                    if record.levelname == "DEBUG"
+                )
+                assert collector._submit_discovery_error_metric.called
+            else:
+                assert any(
+                    "No matching queue of type" in record.message
+                    for record in caplog.records
+                    if record.levelname == "DEBUG"
+                )
+                assert not collector._submit_discovery_error_metric.called
+        else:
+            if auto_discover_queues_via_names:
+                assert any(
+                    "Error inquiring queue names for pattern" in record.message
+                    for record in caplog.records
+                    if record.levelname == "DEBUG"
+                )
+                assert collector._submit_discovery_error_metric.called
+            else:
+                assert any(
+                    "Error discovering queue" in record.message
+                    for record in caplog.records
+                    if record.levelname == "WARNING"
+                )
+                assert not collector._submit_discovery_error_metric.called
+
+
+@pytest.mark.parametrize(
+    "auto_discover_queues_via_names, side_effect_attr",
+    [
+        (False, "MQCMD_INQUIRE_Q"),
+        (True, "MQCMD_INQUIRE_Q_NAMES"),
+    ],
+    ids=["direct_method", "via_names_method"],
+)
+def test_discover_queues_disconnects_on_exception(instance, auto_discover_queues_via_names, side_effect_attr):
+    instance['auto_discover_queues_via_names'] = auto_discover_queues_via_names
+    instance['auto_discover_queues'] = True
+
+    logger = logging.getLogger('test_ibm_mq_disconnect')
+    logger.setLevel(logging.DEBUG)
+
+    collector = make_collector(instance, logger)
+    queue_manager = Mock()
+    pcf_mock = Mock()
+    with patch('datadog_checks.ibm_mq.collectors.queue_metric_collector.pymqi.PCFExecute', return_value=pcf_mock):
+        error = pymqi.MQMIError(2, 9999)
+        setattr(pcf_mock, side_effect_attr, Mock(side_effect=error))
+        collector.discover_queues(queue_manager)
+        assert pcf_mock.disconnect.called
+
+
+@pytest.mark.parametrize(
+    "auto_discover_queues_via_names, patch_method, return_value",
+    [
+        (False, "MQCMD_INQUIRE_Q", []),
+        (True, "MQCMD_INQUIRE_Q_NAMES", [{}]),
+    ],
+    ids=["direct_method", "via_names_method"],
+)
+def test_discover_queues_warns_when_no_queues_found(
+    instance, auto_discover_queues_via_names, patch_method, return_value, caplog
+):
+    instance['auto_discover_queues_via_names'] = auto_discover_queues_via_names
+    instance['auto_discover_queues'] = True
+    instance['queues'] = []
+
+    logger = logging.getLogger('test_ibm_mq_no_queues')
+    logger.setLevel(logging.DEBUG)
```

Comment:
why is this needed?

Reply:
I tried to get this working without it but I can't get the logs to be captured for me to assert. I'll give it another try with the builtin caplog.